pkg/kube/krt/nestedjoinmerge.go

```diff
@@ -0,0 +1,365 @@
+	// ourselves as synced. Do this before returning so collections
+	// aren't added between now and when runQueue is called.
+	subscriptionFunc := func(events []Event[T]) {
+		j.queue.Push(func() error {
+			j.onSubCollectionEventHandler(events)
+			return nil
+		})
+	}
+	reg := j.collections.RegisterBatch(func(e []Event[Collection[T]]) {
+		for _, ev := range e {
+			obj := ev.Latest()
+			switch ev.Event {
+			case controllers.EventAdd:
+				// When a collection is added, subscribe to its events
+				reg := obj.RegisterBatch(subscriptionFunc, true)
+				j.mu.Lock()
+				j.regs[obj.(internalCollection[T]).uid()] = reg
+				j.mu.Unlock()
+			case controllers.EventUpdate:
+				j.handleCollectionUpdate(ev)
+			case controllers.EventDelete:
+				j.handleCollectionDelete(ev)
+			}
+		}
+	}, false)
+	initialCollections := j.collections.List()
+	// Finally, async wait for the primary to be synced. Once it has, we know it has enqueued the initial state.
+	// After this, we can run our queue.
+	// The queue will process the initial state and mark ourselves as synced (from the NewWithSync callback)
+	go j.runQueue(initialCollections, subscriptionFunc, reg)
+
+	return j
+}
+
+func (j *nestedjoinmerge[T]) runQueue(initialCollections []Collection[T], subscriptionFunc func([]Event[T]), reg HandlerRegistration) {
+	// Wait for the container of collections to be synced before we start processing events.
+	if !j.collections.WaitUntilSynced(j.stop) {
+		return
+	}
+	j.mu.Lock()
+
+	// Now that we've subscribed, process the current set of collections.
+	for _, c := range initialCollections {
+		// Save these registrations so we can unsubscribe later if the collection is deleted.
+		// Ensure each sub-collection is synced before we're marked as synced.
+		j.regs[c.(internalCollection[T]).uid()] = c.RegisterBatch(subscriptionFunc, true)
+	}
+
+	regs := append([]HandlerRegistration{reg}, maps.Values(j.regs)...)
+	j.mu.Unlock()
+
+	syncers := slices.Map(regs, func(r HandlerRegistration) cache.InformerSynced {
+		return r.HasSynced
+	})
+
+	if !kube.WaitForCacheSync(j.collectionName, j.stop, syncers...) {
+		return
+	}
+	j.queue.Run(j.stop)
+}
+
+func (j *nestedjoinmerge[T]) handleCollectionUpdate(e Event[Collection[T]]) {
+	innerCollection := e.Latest().(internalCollection[T])
+	log.Debugf("NestedJoinWithMergeCollection: Collection %s (uid %s) updated, recalculating merged values", innerCollection.name(), innerCollection.uid())
+	// Get all of the elements in the old collection
+	oldCollectionValue := *e.Old
+	newCollectionValue := *e.New
+	// Wait for the new collection to be synced before we process the update.
+	if !newCollectionValue.WaitUntilSynced(j.stop) {
+		log.Warnf("NestedJoinWithMergeCollection: Collection %s not synced, skipping update event", newCollectionValue.(internalCollection[T]).uid())
+	}
+	// Stop the world and update our outputs with new state for everything in the collection.
+	j.mu.Lock()
```

Comment:
Should this locking happen before or after we wait for sync? Is it possible that we have something like:

1. Collection is updated from v1 -> v2, so we start waiting
2. While we are waiting, we do v2 -> v3
3. Now we hit this same path for v3, whichever readies first between v2 and v3 will go first 

Main concerns would be v3 syncing, v2 is no longer really relevant, but it loses the race and we process it's (probably empty?) items last. 

I think either we lock before we wait _or_ we have an additional check to see if there was another collectionUpdate while we were waiting for sync. 

Probably a very rare case in practice..

Reply:
The `RegisterBatch` call on the collection of collections happens once in one goroutine, so 3) would never happen right? The event processor would block until the v1 -> v2 event had been processed by this handler. Then we'd get the v2 -> v3 event